sdk/src/main/scala/com/horizen/forge/Forger.scala

```diff
@@ -47,16 +47,28 @@ class Forger(settings: SidechainSettings,
     Await.result(future, timeoutDuration).asInstanceOf[View]
   }
 
-  protected def getNextBlockForgingInfo(): Try[ForgingInfo] = Try {
+  protected def getNextBlockForgingInfo: Try[ForgingInfo] = Try {
     val view = getNodeView
     val parentId: Block.BlockId = view.history.bestBlockId
     val timestamp: Long = Instant.now.getEpochSecond
-    val mainchainBlockRefToInclude: Seq[MainchainBlockReference] =
-      mainchainSynchronizer.getNewMainchainBlockReferences(view.history, SidechainBlock.MAX_MC_BLOCKS_NUMBER)
+
+    var withdrawalEpochMcBlocksLeft = params.withdrawalEpochLength - view.history.bestBlockInfo.withdrawalEpochInfo.index
+    if(withdrawalEpochMcBlocksLeft == 0) // current best block is the last block of the epoch
+      withdrawalEpochMcBlocksLeft = params.withdrawalEpochLength
+
+    val mainchainBlockRefToInclude: Seq[MainchainBlockReference] = mainchainSynchronizer.getNewMainchainBlockReferences(
+      view.history,
+      Math.min(SidechainBlock.MAX_MC_BLOCKS_NUMBER, withdrawalEpochMcBlocksLeft) // to not to include mcblock references from different withdrawal epochs
+    )
+
     val txsToInclude: Seq[SidechainTransaction[Proposition, NoncedBox[Proposition]]] =
-      view.pool.take(SidechainBlock.MAX_SIDECHAIN_TXS_NUMBER) // TO DO: problems with types
-        .map(t => t.asInstanceOf[SidechainTransaction[Proposition, NoncedBox[Proposition]]])
-        .toSeq
+      if(mainchainBlockRefToInclude.size == withdrawalEpochMcBlocksLeft) { // SC block is going to become the last block of the withdrawal epoch
+        Seq() // no SC Txs allowed
+      } else { // SC block is in the middle of the epoch
+        view.pool.take(SidechainBlock.MAX_SIDECHAIN_TXS_NUMBER) // TO DO: problems with types
+          .map(t => t.asInstanceOf[SidechainTransaction[Proposition, NoncedBox[Proposition]]])
+          .toSeq
```

Comment:
view.pool.take(SidechainBlock.MAX_SIDECHAIN_TXS_NUMBER) // TO DO: problems with types
          .map(t => t.asInstanceOf[SidechainTransaction[Proposition, NoncedBox[Proposition]]])(collections.breakOut) coud be used ))

Reply:
As for me `...(collections.breakOut)` is too scala